client/mixnet/src/request.rs

```diff
@@ -0,0 +1,109 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::{config::SubstrateConfig, error::Error};
+use blake2::{
+	digest::{consts::U16, Mac},
+	Blake2bMac,
+};
+use codec::{Decode, DecodeAll};
+use futures::channel::oneshot;
+use log::warn;
+use mixnet::core::{Delay, MessageId, PostErr, Scattered};
+use sp_core::Bytes;
+use std::time::Duration;
+
+const LOG_TARGET: &str = "mixnet";
+
+fn send_err<T>(reply_sender: oneshot::Sender<Result<T, Error>>, err: Error) {
+	if let Err(Err(err)) = reply_sender.send(Err(err)) {
+		warn!(target: LOG_TARGET, "Failed to inform requester of error: {err}");
+	}
+}
+
+fn send_reply<T: Decode>(reply_sender: oneshot::Sender<Result<T, Error>>, mut data: &[u8]) {
+	let res = match Result::decode_all(&mut data) {
+		Ok(res) => res.map_err(Error::Remote),
+		Err(_) => Err(Error::BadReply),
+	};
+	match reply_sender.send(res) {
+		Ok(_) => (),
+		Err(Ok(_)) => warn!(target: LOG_TARGET, "Failed to send reply to requester"),
+		Err(Err(err)) => warn!(target: LOG_TARGET, "Failed to inform requester of error: {err}"),
+	}
+}
+
+pub const SUBMIT_EXTRINSIC: u8 = 0;
+
+const EXTRINSIC_DELAY_PERSONA: &[u8; 16] = b"submit-extrn-dly";
+
+pub fn extrinsic_delay(message_id: &MessageId, config: &SubstrateConfig) -> Duration {
+	let h = Blake2bMac::<U16>::new_with_salt_and_personal(message_id, b"", EXTRINSIC_DELAY_PERSONA)
+		.expect("Key, salt, and persona sizes are fixed and small enough");
```

Comment:
Both the sender and receiver calculate this; the sender uses it to determine the timeout. Only the sender and receiver should know the message ID, so I think this is fine. An alternative would be to just randomly pick a delay in the receiver and have the sender use the 99.9th percentile for the timeout. I think having the sender know the exact delay is preferable as it allows for tighter timeouts, but I could probably be convinced otherwise!

Reply:
Cannot really say, but sure is something to keep in mind (should be in the top level documentation).
Is the message ID the same for every hops?